Plugins/SwiftLintPlugin/SwiftLintPlugin.swift

```diff
@@ -3,81 +3,138 @@ import PackagePlugin
-                executable: tool.path,
-                arguments: arguments,
-                outputFilesDirectory: outputFilesDirectory
-            )
+                executable: executable.path,
+                arguments: arguments + cacheArguments + swiftFiles.map(\.string),
+                environment: environment,
+                outputFilesDirectory: outputPath)
         ]
     }
 }
 
 #if canImport(XcodeProjectPlugin)
+
 import XcodeProjectPlugin
 
+// swiftlint:disable:next no_grouping_extension
 extension SwiftLintPlugin: XcodeBuildToolPlugin {
-    func createBuildCommands(context: XcodePluginContext, target: XcodeTarget) throws -> [Command] {
-        let inputFilePaths = target.inputFiles
+    func createBuildCommands(
+        context: XcodePluginContext,
+        target: XcodeTarget
+    ) throws -> [Command] {
+        try makeCommand(executable: context.tool(named: "swiftlint"),
+                        swiftFiles: swiftFiles(target: target),
+                        environment: environment(context: context, target: target),
+                        pluginWorkDirectory: context.pluginWorkDirectory)
+    }
+
+    /// Collects the paths of the Swift files to be linted.
+    private func swiftFiles(target: XcodeTarget) -> [Path] {
+        target
+            .inputFiles
             .filter { $0.type == .source && $0.path.extension == "swift" }
             .map(\.path)
-        return createBuildCommands(
-            inputFiles: inputFilePaths,
-            packageDirectory: context.xcodeProject.directory,
-            workingDirectory: context.pluginWorkDirectory,
-            tool: try context.tool(named: "swiftlint")
-        )
+    }
+
+    /// Creates an environment dictionary containing a value for the `BUILD_WORKSPACE_DIRECTORY` key.
+    ///
+    /// This method locates the topmost `.swiftlint.yml` config file within the project directory for this target's
+    /// Swift source files and sets the config file's containing directory as the `BUILD_WORKSPACE_DIRECTORY` value.
+    /// The project directory is used if a config file is not found.
+    ///
+    /// The `BUILD_WORKSPACE_DIRECTORY` environment variable controls SwiftLint's working directory.
+    ///
+    /// Reference: [https://github.com/realm/SwiftLint/blob/0.50.3/Source/swiftlint/Commands/SwiftLint.swift#L7](
+    /// https://github.com/realm/SwiftLint/blob/0.50.3/Source/swiftlint/Commands/SwiftLint.swift#L7
+    /// )
+    private func environment(
+        context: XcodePluginContext,
+        target: XcodeTarget
+    ) throws -> [String: String] {
+        let projectDirectory: Path = context.xcodeProject.directory
+        let swiftFiles: [Path] = swiftFiles(target: target)
+        let swiftFilesNotInProjectDirectory: [Path] = swiftFiles.filter { !$0.isDescendant(of: projectDirectory) }
+
+        guard swiftFilesNotInProjectDirectory.isEmpty
+        else { throw SwiftLintPluginError.swiftFilesNotInProjectDirectory(projectDirectory) }
```

Comment:
This style is uncommon for this project.
```suggestion
        guard swiftFilesNotInProjectDirectory.isEmpty else { 
            throw SwiftLintPluginError.swiftFilesNotInProjectDirectory(projectDirectory)
        }
```